java/ql/src/semmle/code/java/Annotation.qll

```diff
@@ -44,12 +44,65 @@ class Annotation extends @annotation, Expr {
     result = this.getType().getAnnotationElement(name)
   }
 
-  /** Gets a value of an annotation element. */
+  /**
+   * Gets a value of an annotation element. This includes default values in case
+   * no explicit value is specified.
+   */
   Expr getAValue() { filteredAnnotValue(this, _, result) }
 
-  /** Gets the value of the annotation element with the specified `name`. */
+  /**
+   * Gets the value of the annotation element with the specified `name`.
+   * This includes default values in case no explicit value is specified.
+   */
   Expr getValue(string name) { filteredAnnotValue(this, this.getAnnotationElement(name), result) }
 
+  /**
+   * If the value type of the annotation element with the specified `name` is an enum type,
+   * gets the enum constant used as value for that element. This includes default values in
+   * case no explicit value is specified.
+   */
+  EnumConstant getValueEnumConstant(string name) { result = getValue(name).(FieldRead).getField() }
```

Comment:
This isn't really specific to annotations. Perhaps provide a class `EnumConstantRead extends FieldRead` with relevant convenience methods?

Reply:
> This isn't really specific to annotations.

It is, isn't it? Enums are one of the value types an annotation type element can have.
Consider this example for [`@Target`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/annotation/Target.html):
```java
@Target(ElementType.ANNOTATION_TYPE)
public @interface MetaAnnotationType {
    ...
}
```
The added CodeQL predicate would allow easily getting `ElementType.ANNOTATION_TYPE` here.

How would a `EnumConstantRead` help for this use case? Or do you mean something different?

Reply:
I mean that the desire to interpret a `FieldRead` as an enum constant read goes far beyond annotations, so we should provide something like

```
class EnumConstantRead extends FieldRead {

  EnumConstantRead() { this.getField() instanceof EnumConstant }

  EnumConstant getEnumConstant() { result = this.getField() }

}
```

Then use it in this method.

Reply:
Ah, ok I thought you meant this predicate should be removed.

Though I am not sure if `EnumConstantRead` adds that much value; it probably only saves you one `EnumConstant` cast for the result of `getField()` in most cases.